authnz/src/main/java/io/stargate/auth/AuthenticationService.java

```diff
@@ -16,14 +16,17 @@
 package io.stargate.auth;
 
 import io.stargate.db.Authenticator.SaslNegotiator;
+import java.util.Map;
 
 public interface AuthenticationService {
 
-  String createToken(String key, String secret) throws UnauthorizedException;
+  String createToken(String key, String secret, Map<String, String> headers)
```

Comment:
I'd rename `heads` to `properties`, for the sake of clarify since they do not always come from HTTP header, nor are they required to be HTTP headers, AFAIK.

Reply:
Suggestion: it might be worth defining a separate holder class for "headers" and also make it encapsulate the `key` and the `secret` WDYT?

Reply:
Another point of having a custom "parameters" or perhaps "context" object: Currently there are two distinct use cases: 1) for HTTP-based APIs 2) for CQL. The latter uses the "source host name" from `ClientInfo` for defining the tenant. So, rather than converting everything to the headers map, we could have both pieces of data defined independently in the "context" object. The auth implementation will use one of them (or both) as appropriate. The benefit would be clearer separation of concerns, e.g. "tenant ID" would not have be an common entity. Only those auth implementations that have that notion will have code dealing with it. Other implementation will have to be aware of that concept.

Reply:
Similarly, `AuthenticationSubject` would only contain a tenant "Principal" when tenants are in use. Other authorization implementations will not need to be aware of tenants. At java API level we'd have a generic collection of "Principal" objects.

Reply:
We were discussing that with @dougwettlaufer and it was suggested to pass the map of headers because this mechanism is generic:
> I'd rename headers to properties, for the sake of clarify since they do not always come from HTTP header, nor are they required to be HTTP headers, AFAIK.

The same comment, we decided to stick to the headers. The CQL use case is a special case and we can make it generic by using the generic map I think. 

>Suggestion: it might be worth defining a separate holder class for "headers" and also make it encapsulate the key and the secret WDYT?

both `key` and `secret` are not related to changes in this PR. I would like to reduce the number of refactoring changes here; we can address is separate.

Reply:
Ultimately, headers are required for authentication, right? The result of a successful authentication is an `AuthenticationSubject`, so why not record the headers inside `AuthenticationSubject` for later use during authorization? I believe this would be consistent with our auth model and might also require less code changes.

Regarding "principals" that is a secondary point. I'd be ok with using "headers", but I'm not sure it's worth putting them as separate parameters into authorization API methods :thinking:

Reply:
I am +1 on trying that path and compare to the current design. However, I feel that because of the timeline we should do it after the merge and initial testing. Feel free to create a follow-up ticket.